crates/turbo-tasks-memory/src/edges_set.rs

```diff
@@ -0,0 +1,592 @@
+use std::{hash::BuildHasherDefault, mem::replace};
+
+use auto_hash_map::{map::Entry, AutoMap, AutoSet};
+use rustc_hash::FxHasher;
+use smallvec::SmallVec;
+use turbo_tasks::{CellId, TaskId, TraitTypeId, ValueTypeId};
+
+use crate::multi_iter::IntoIters4;
+
+#[derive(Hash, Copy, Clone, PartialEq, Eq)]
+pub enum TaskEdge {
+    Output(TaskId),
+    Cell(TaskId, CellId),
+    Collectibles(TaskId, TraitTypeId),
+    Child(TaskId),
+}
+
+impl TaskEdge {
+    fn task_and_edge_entry(self) -> (TaskId, EdgeEntry) {
+        match self {
+            TaskEdge::Output(task) => (task, EdgeEntry::Output),
+            TaskEdge::Cell(task, cell_id) => (task, EdgeEntry::Cell(cell_id)),
+            TaskEdge::Collectibles(task, trait_type_id) => {
+                (task, EdgeEntry::Collectibles(trait_type_id))
+            }
+            TaskEdge::Child(task) => (task, EdgeEntry::Child),
+        }
+    }
+}
+
+#[derive(Hash, Copy, Clone, PartialEq, Eq)]
+enum EdgeEntry {
+    Output,
+    Child,
+    Cell(CellId),
+    Collectibles(TraitTypeId),
+}
+
+impl EdgeEntry {
+    fn into_dependency(self, task: TaskId) -> TaskEdge {
+        match self {
+            EdgeEntry::Output => TaskEdge::Output(task),
+            EdgeEntry::Cell(cell_id) => TaskEdge::Cell(task, cell_id),
+            EdgeEntry::Collectibles(trait_type_id) => TaskEdge::Collectibles(task, trait_type_id),
+            EdgeEntry::Child => TaskEdge::Child(task),
+        }
+    }
+}
+
+type ComplexSet = AutoSet<EdgeEntry, BuildHasherDefault<FxHasher>, 3>;
+
+enum EdgesEntry {
+    Empty,
+    Output,
+    Child,
+    ChildAndOutput,
+    Cell0(ValueTypeId),
+    ChildAndCell0(ValueTypeId),
+    OutputAndCell0(ValueTypeId),
+    ChildOutputAndCell0(ValueTypeId),
+    Complex(Box<ComplexSet>),
+}
+
+impl EdgesEntry {
+    fn from(entry: EdgeEntry) -> Self {
+        match entry {
+            EdgeEntry::Output => EdgesEntry::Output,
+            EdgeEntry::Child => EdgesEntry::Child,
+            EdgeEntry::Cell(CellId { type_id, index }) => {
+                if index == 0 {
+                    EdgesEntry::Cell0(type_id)
+                } else {
+                    let mut set = AutoSet::default();
+                    set.insert(EdgeEntry::Cell(CellId { type_id, index }));
+                    EdgesEntry::Complex(Box::new(set))
+                }
+            }
+            EdgeEntry::Collectibles(trait_type_id) => {
+                let mut set = AutoSet::default();
+                set.insert(EdgeEntry::Collectibles(trait_type_id));
+                EdgesEntry::Complex(Box::new(set))
+            }
+        }
+    }
+
+    fn len(&self) -> usize {
+        match self {
+            EdgesEntry::Empty => unreachable!(),
```

Comment:
While this may not happen in practice, I think it would still be reasonable to implement this as 0.

Reply:
`EdgesEntry::Empty` indicates a bug in a previous operations. I rather like it to panic early.